File: src/org/espenhahn/allocate/likepaxos/registry/debug/mvc/objects/Drawable.java
package org.espenhahn.allocate.likepaxos.registry.debug.mvc.objects;

import java.awt.Graphics2D;
import java.util.Map.Entry;
import java.util.SortedMap;
import java.util.TreeMap;

public abstract class Drawable {
	
	private SortedMap<Integer, Text> labels;
	private Drawable owner;
	
	public Drawable() {
		this(null);
	}
	
	public Drawable(Drawable owner) {
		this.labels = new TreeMap<>();
		this.owner = owner;
	}
	
	protected void setOwner(Drawable d) {
		this.owner = d;
	}
	
	public void addLabel(int id, Text label) {
		labels.put(id, label);
		label.setOwner(this);
	}
	
	public void renderLabels(Graphics2D g) {
		for (Entry<Integer, Text> e: labels.entrySet()) {
			e.getValue().render(g);
		}
	}
	
	public abstract void render(Graphics2D g2);
	
	public int getX() {
		return getRelativeX() + (owner != null ? owner.getX() : 0);
	}
	
	public abstract int getRelativeX();
	
	public int getY() {
		return getRelativeY() + (owner != null ? owner.getY() : 0);
	}
	
	public abstract int getRelativeY();
	
	public abstract void setRelativePos(int x, int y);
	
	public void translate(int dx, int dy) {
		setRelativePos(getRelativeX() + dx, getRelativeY() + dy);
	}
}

File: src/org/espenhahn/allocate/likepaxos/registry/debug/mvc/objects/Drawable.kt
package org.espenhahn.allocate.likepaxos.registry.debug.mvc.objects

import java.awt.Graphics2D
import java.util.*

abstract class Drawable @JvmOverloads constructor(relativeX: Int, relativeY: Int, private var owner: Drawable? = null) {

    private val labels: SortedMap<Int, Text>

    var relativeX: Int
        private set

    var relativeY: Int
        private  set

    init {
        this.relativeX = relativeX
        this.relativeY = relativeY
        this.labels = TreeMap<Int, Text>()
    }

    protected fun setOwner(d: Drawable) {
        this.owner = d
    }

    fun addLabel(id: Int, label: Text) {
        labels.put(id, label)
        label.setOwner(this)
    }

    fun renderLabels(g: Graphics2D) {
        for ((_, value) in labels) {
            value.render(g)
        }
    }

    abstract fun render(g2: Graphics2D)

    fun setRelativePos(x: Int, y: Int) {
        this.relativeX = x
        this.relativeY = y
    }

    val x: Int
        get() = relativeX + (owner?.x ?: 0)

    val y: Int
        get() = relativeY + (owner?.y ?: 0)

    open fun translate(dx: Int, dy: Int) {
        setRelativePos(relativeX + dx, relativeY + dy)
    }
}
